### ISSUE TITLE
issue with gzip and changeOrigin: true

### ISSUE THREAD

**Issue by requiredfield** [nodejitsu/node-http-proxy#795]
When the `target` of the following node-http-proxy server points to an origin that serves gzipped responses to clients sending `Accept-Encoding: gzip`, node-http-proxy prepends garbage bytes to the response causing decoding errors:

```js
var httpProxy = require('http-proxy');
httpProxy.createServer({
  target: '$origin',
  changeOrigin: true,
  xfwd: true
}).listen(8000);
```

When `changeOrigin: true` is removed, the problem goes away, but we rely on keeping it there.

Can you confirm that this is a bug in node-http-proxy, and if so, give any pointers on how to fix it?

Thanks for taking a look and for the great work on node-http-proxy.

**Comment by requiredfield** [comment]
In case anyone else is hitting this and is looking for a workaround, stripping request's `Accept-Encoding` header effectively disables gzip on the origin, avoiding the bug:
```js
var httpProxy = require('http-proxy');
var proxy = httpProxy.createServer({
  target: '$origin',
  changeOrigin: true,
  xfwd: true
});
proxy.on('error', function (e) { console.error(e); });

var http = require('http');
var server = http.createServer(function (req, res) {
  delete req.headers['accept-encoding'];
  proxy.web(req, res);
});
var port = process.env.port || 8000;
console.info('listening on port', port);
server.listen(port);
```

**Comment by jcrugzz** [comment]
@requiredfield This is fascinating that this is only when `changeOrigin: true` option is used. Could you submit your example as a failing test so we can get started in finding the proper fix?

**Comment by Behemouth** [comment]
@jcrugzz  This only happens to some of our sites ,e.g.   http://fw.m.pingfan.pw/ , only when set changeOrigin to true and enabled gzip encoding. The gzipped output body has been prepended  three  garbage bytes e.g. "22f" and "\r\n",total five bytes,that caused decoding error.If I strip these five bytes then it can be decoded correctly but missed the last five bytes,usually is "html>" .

PS: I've found `xfwd` option also affects the gzip output. When switch xfwd between true and false,the three garbage bytes varies.

**Comment by jcrugzz** [comment]
@Behemouth hmm it has to be because the headers are changing, its not able to correctly gzip the entire request. This is something that I would definitely need to dig deeper into but would love some help. What node/io.js version are you using?

**Comment by requiredfield** [comment]
Using latest release of Node (0.12.0) and latest release of node-http-proxy. Would it help if we tried older versions?

**Comment by jcrugzz** [comment]
@requiredfield yes I'd be interested to see if it represents the same behavior in `0.10.37` and I would also test [`io.js@1.6.1`](https://iojs.org/en/index.html) if you have the chance. I just know that node `0.12.0` has bugs so I want to eliminate that as a variable.

**Comment by Behemouth** [comment]
@jcrugzz  

My Node.js version has this bug too:
```
$ nodejs --version
v0.10.25
$ npm list
└─┬ http-proxy@1.9.0
     ├── eventemitter3@0.1.6
     └── requires-port@0.0.0
```

**Comment by mclausen** [comment]
Hi Nodejitsu, 
We currently hit the same problem, and removing the header seems like the only workaround we could think of. Did you guys happen to get around this issue other more elegant way?

In addition, but no directly related we had to strip the compression middleware app.use(compression()) from server.js to keep expression.js from tampering with our requests.
